In [10]:
from datasets import load_dataset

# Specify the subsets to load
subsets = [
    "Bible",
    "TED2020",
    "Opensubtitles",
    "S2ORC",
    "CodePileReddit2022",
    "USENET",
    "PileV2Reddit2020",
    "arXiv",
    "CodePileReddit2020",
    "CodePileReddit2021"
]

# Load each subset and store it in a dictionary
subset_datasets = {}
for subset in subsets:
    subset_dataset = load_dataset("CarperAI/pile-v2-small-filtered", data_dir=f"data/{subset}")
    subset_datasets[subset] = subset_dataset

# Print information about each subset dataset
for subset, dataset in subset_datasets.items():
    print(f"Subset: {subset}")
    print(dataset)
    print()


Subset: Bible
DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'meta'],
        num_rows: 1
    })
})

Subset: TED2020
DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'meta'],
        num_rows: 1000
    })
})

Subset: Opensubtitles
DatasetDict({
    train: Dataset({
        features: ['text', 'meta', 'id', 'check_char_repetition_criteria', 'check_flagged_words_criteria', 'check_stop_word_ratio_criteria'],
        num_rows: 1000
    })
})

Subset: S2ORC
DatasetDict({
    train: Dataset({
        features: ['text', 'meta', 'id'],
        num_rows: 1000
    })
})

Subset: CodePileReddit2022
DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'meta'],
        num_rows: 1000
    })
})

Subset: USENET
DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'meta'],
        num_rows: 1000
    })
})

Subset: PileV2Reddit2020
DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'meta', 'check_char_repetition_crite

The Pile is an 800 GB corpus that enhances a model’s generalization capability across a broader context. It was curated from 22 diverse datasets, mostly from academic or professional sources. The Pile was instrumental in training various LLMs, including GPT-Neo, LLaMA, and OPT.
A small subset in each dataset of pile-v2 of pile-v2 dataset, each has 1,000 random samples from the original dataset. The dataset has 255MB of text.


In [8]:
from genbit.genbit_metrics import GenBitMetrics

lm_dataset = []
for split in subset_dataset.keys():
    lm_dataset.extend(subset_dataset[split]['text'])
    
genbit_metrics_object2 = GenBitMetrics(language_code="en", context_window=5, distance_weight=0.95, percentile_cutoff=80)

print(lm_dataset)

genbit_metrics_object2.add_data(lm_dataset, tokenized=False)

# To generate the gender bias metrics, we run `get_metrics` by setting `output_statistics` and `output_word_lists` to false, we can reduce the number of metrics created.

metrics = genbit_metrics_object2.get_metrics(output_statistics=True, output_word_list=True)


["In 2021, the conversation below took place in an online community called Android:\nssource1: Out of curiosity, what's everybody storing on their phones? \n\nI'm still on a S8 with 64gb and got a 64gb microSD while back - it's been plenty for me. In terms of files, it's mostly photos, some 1080p60 video recordings, offline music synced from YT Music, and random documents.\nSowers25: My music library on my phone is currently 135gb. Now to be fair, 95% of it 24 bit FLAC, so the files are huge. But still. I also have 50gb of movies and 120gb of tv shows.\nCry_Wolff: &gt; I also have 50gb of movies and 120gb of tv shows.\n\nPeople really do watch movies and show on their phones huh.\npitchbend: No I use an HDMI cable to watch them at hotels friend houses etc or cast tayloredward@example.org. It's nice to have them offline and always ready with you.\n\n\n", "In 2021, the conversation below took place in an online community called MacOS:\nLiLHyena: Im not getting this update, i had the 11.2

In [9]:
print(metrics)

{'genbit_score': 0.5411597028875228, 'percentage_of_female_gender_definition_words': 0.04847207586933614, 'percentage_of_male_gender_definition_words': 0.1787846856340007, 'percentage_of_non_binary_gender_definition_words': 0.7727432384966632, 'percentage_of_trans_gender_definition_words': 1.0, 'percentage_of_cis_gender_definition_words': 0.0, 'additional_metrics': {'avg_bias_ratio': 0.4210452786193743, 'avg_bias_conditional': 0.38803075735095605, 'avg_bias_ratio_absolute': 0.5407671950213045, 'avg_bias_conditional_absolute': 0.5411597028875228, 'avg_non_binary_bias_ratio': -0.9637705252276646, 'avg_non_binary_bias_conditional': -0.8377377705691156, 'avg_non_binary_bias_ratio_absolute': 1.1459020022162039, 'avg_non_binary_bias_conditional_absolute': 1.0599583482581747, 'avg_trans_cis_bias_ratio': -1.6457351389547532, 'avg_trans_cis_bias_conditional': -1.3070403886719204, 'avg_trans_cis_bias_ratio_absolute': 1.6457351389547532, 'avg_trans_cis_bias_conditional_absolute': 1.32755777801803

In this case, the GenBit score is 0.54, indicating a moderate bias towards male gender words.